### PR TITLE
Add unit command bar to TD.

### PR DESCRIPTION
**Pull request by pchote**
This implements the backend for #3789, and the frontend for TD.  The other mods will come in future PRs after this is merged.

unitcommand

The bar exposes clickable UI for all the keyboard-only actions that were previously hacked into `UnitCommandWidget` (now deprecated and will be removed once the bar is implemented in the rest of the mods).

This is important for discoverability and usability / accessibility by players who may not be able to use the keyboard.  For this reason, I consider it to be an essential part of the unit stance/targeting rework that we are aiming for in the next playtest.

### CODE REVIEW

**Review by cjshmyr — APPROVED** [review]

**Comment by cjshmyr** [comment]
:+1: -- edge scrolling overlaps half of the small buttons but pretty minor

**Comment by Micr0Bit** [comment]
>:+1: -- edge scrolling overlaps half of the small buttons but pretty minor

i second that 

:+1:

**Review by Micr0Bit — APPROVED** [review]
>:+1: -- edge scrolling overlaps half of the small buttons but pretty minor

i second that

**Comment by pchote** [comment]
Added a commit to fix the edge scrolling, but this then of course means that you lose the ability to edge scroll on the bottom left part of the screen.

I could move the bar up so it is floating with a ~10px gap, like the sidebars, to allow scrolling, but we then lose the benefit of fitt's law for the actions and this forces the chat and debug graphs up too.

**Comment by pchote** [comment]
Fixed the edge scrolling with a compromise:
* Bar is now floating with a 5px gap between left and bottom edges (bad for fitt's law, but consistent with what C&C3 did).
* Edge scrolling threshold reduced to 5px, and cursor hotspots fixed on TD/RA/TS (D2K was already fine) so that cursors show properly without needing a huge threshold.

@cjshmyr and @Micr0Bit, can you confirm you :+1:s still hold with these changes?

**Comment by cjshmyr** [comment]
:+1:

**Comment by Micr0Bit** [comment]
:+1:

File: OpenRA.Mods.Common/Widgets/CommandBarButtonWidget.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		public Func<string> GetTooltipDesc;
+		public Func<string> GetKeyName;
+
+		[ObjectCreator.UseCtor]
+		public CommandBarButtonWidget(ModData modData)
+			: base(modData)
+		{
+			GetTooltipDesc = () => TooltipDesc;
+			GetKeyName = () => null;
+		}
```
Comment (rob-v) [inline OpenRA.Mods.Common/Widgets/CommandBarButtonWidget.cs]: You could add: OnKeyPress = ki => { if (ki.Event == KeyInputEvent.Down && !ki.IsRepeat) OnClick();
You can compare/hear the difference by holding A in release and this PR.

my suggestion from another PR: change default ButtonWidget.OnKeyPress from { OnClick(); } to {ki => { if (ki.Event == KeyInputEvent.Down && !ki.IsRepeat) OnClick(); }

**Review by rob-v — COMMENTED** [review]
Works great, few comments.

It is intentional that you press A key to turn 'Attack move' on once, but you need to hold Alt key, right?
You can turn e.g. Attack move and Force move by commandbar left click, but you can't turn it off with next left click. Left clicks turns on both in the same way, but key press works differently (1 press vs holding).

**Comment by pchote** [comment]
Thanks for the excellent feedback @rob-v.  I have quite significantly reworked the code to take your suggestions, and my own additions based on those, into account.  Needs a rereview.

**Comment by pchote** [comment]
Updated screenshot showing the edge scrolling margin and the new force-modifier tooltip style:
screen shot 2017-06-04 at 14 13 23

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/CommandBarLogic.cs [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+			{
+				BindButtonIcon(deployButton);
+
+				deployButton.GetKey = _ => ks.DeployKey;
+				deployButton.IsDisabled = () => { UpdateStateIfNecessary(); return !selectedDeploys.Any(Exts.IsTraitEnabled); };
+				deployButton.IsHighlighted = () => deployHighlighted > 0;
+				deployButton.OnKeyPress = ki =>
+				{
+					if (ki.Event == KeyInputEvent.Down && !ki.IsRepeat)
+						deployButton.OnClick();
```
Comment (rob-v) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/CommandBarLogic.cs]: This is now unnecessary with DisableKeyRepeat?

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/CommandBarLogic.cs [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+				};
+
+				deployButton.OnClick = () =>
+				{
+					deployHighlighted = 2;
```
Comment (rob-v) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/CommandBarLogic.cs]: I noticed that there is additonal blick due to this on mouse click. The button is highlighted already on button click. This should be moved to OnKeyPress (also for Scatter, Stop).

**Review by rob-v — COMMENTED** [review]
I would feel better if you would keep the modifiers not highlighted due to required changes in new ButtonTooltipWithDescHighlightLogic or at least if it could be unified with also extended ButtonTooltipLogic. You even added highlighting feedback on hotkey press.

Last small issues/questions.

**Comment by pchote** [comment]
> or at least if it could be unified with also extended ButtonTooltipLogic. 

Of the three options:
1. Call a virtual method from the base class's constructor
2. Deviate from our standard code patterns for UI logic classes
3. Duplicate ~50 lines of code.

Option 3 is the least bad.

**Comment by pchote** [comment]
Fixed the other two issues.

**Review by rob-v — APPROVED** [review]
It is really great. 
👍 (especially for DisableKeyRepeat)

**Comment by cjshmyr** [comment]
I've encountered starting unit selections (not control groups) being in a state where there's no command bar options. I've had it happen with:

Starting a game as Nod and selecting all the starting infantry units
Starting a game as GDI and selecting all starting units

image

**Comment by GraionDilach** [comment]
Shouldn't this fix #13105?

**Comment by pchote** [comment]
> I've encountered starting unit selections being in a state where there's no command bar options.

I couldn't reproduce this (it was a problem in the first version of the PR, but I thought I had fixed it), but just to be safe I replaced the hash with a simple and foolproof counter.

> Shouldn't this fix #13105?

Not this pr, but I will get there in the RA pr.

**Comment by cjshmyr** [comment]
Thanks, that worked 👍. @rob-v: anything else?

**Comment by rob-v** [comment]
👍

**Comment by cjshmyr** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/244d86d05a79da2ce6204e5c40c6ddb53e191209...e7618f5d466ed33496075d60ed9f35e82e37e9af), probably want to move that into general improvements once it hits all mods.

**Pull request merged**